pkg/services/sqlstore/migrations/ualert/tables.go

```diff
@@ -433,3 +455,81 @@ func AddAlertImageMigrations(mg *migrator.Migrator) {
 		Postgres("ALTER TABLE alert_image ALTER COLUMN url TYPE VARCHAR(2048);").
 		Mysql("ALTER TABLE alert_image MODIFY url VARCHAR(2048) NOT NULL;"))
 }
+
+func extractAlertmanagerConfigurationHistoryMigration(mg *migrator.Migrator) {
```

Comment:
Yeah, good idea, `ualert.go` was not the best place.

Not for this PR, but I'm just curious - WDYT about eventually having one file for each non-trivial migration? The legacy->UA migration could be its own subpackage, etc. Or is that overkill?

Reply:
I think it still wouldn't solve the underlying issue. IMO, the confusion is mostly caused by the fact that our main legacy -> new alerting migration isn't really a migration:

- It's mutable. It changes quite often.
- It's reversible. This means, it needs to be constantly updated to work with the most recent data model otherwise there can be issues if the the migration is reverted and reapplied.
- It's optional. So, knowing it was applied gives us no information on what their migrated alerts look like now since it depends entirely on the version they did the migration and if they have since modified them.
- Because of the above, it can't be composed in a way to apply step-wise over time; the migration is all or nothing. This means that if someone has previously migrated in, say 8.4, they would have to revert and remigrate to take advantage of any improvements since then.

These migration assumptions more closely resemble an 'upgrade tool/cli' and if I were to recommend making large changes to it now that's probably what I would go for. Basically, extract the legacy migration that actually converts their alerting data as a tool and keep the other migrations (such as table schemas) as ones that are immutable and can be reasoned around like any other migration in Grafana.

This won't change any of the points from above. We'd still have to maintain the tool to upgrade to the current data model. But, it would mean our actual migrations can be reasoned around more clearly.